[PKM-30] feat: Go router setup

yohancai opened the pull request:
- Tentative favorite UI view
- set up of go router
- created home_page to facilitate go router
-
https://user-images.githubusercontent.com/58543326/181441002-1469d890-2436-4249-8ea1-d0ff5920f37b.mp4

Commit:
[PKM-30] feat: Go router setup

- initial set up of go router
- initial UI of favorite page
- created home page to facilitate go router

Commit:
[PKM-30] feat: Go router setup

- Tentative favorite UI view
- set up of go router
- created home_page to facilitate go router

Commit:
[PKM-30] feat: Go router setup

- Tentative favorite UI view
- set up of go router
- created home_page to facilitate go router

Commit:
[PKM-30] feat: Go router setup

- Tentative favorite UI view
- set up of go router
- created home_page to facilitate go router

Commit:
[PKM-30] feat: Go router setup

- Tentative favorite UI view
- set up of go router
- created home_page to facilitate go router

JotamSleonardo commented on lib/features/favorite_list/favorite_pokemon.dart:
```diff
@@ -0,0 +1,12 @@
+  @override
+  Widget build(BuildContext context) {
+    return const Center(
+      child: Text('Favorites'),
```
Move it to constant

JotamSleonardo commented on lib/utils/generate_pokemon_image.dart:
```diff
@@ -0,0 +1,4 @@
+String generateImage(int index) {
+  index += 1;
+  return "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/$index.png";
```
This is what I told you last time about putting this into constants so that you can just use it like 
k.pokeImage.replase(with index)

yohancai replied:
Okay I understand now hehe

Commit:
[PKM-30] feat: Go router setup

- fixed generate Image

Commit:
[PKM-30] feat: Go router Setup

- initial set up of go router
- initial nav bar for home page

Commit:
[PKM-30] feat: Go router setup

- added route and name to connector
- created initial favorite connector and vm
- created initial pokemon connector and vm

Commit:
[PKM-30] feat: Go router setup

- lints fixed

JotamSleonardo commented:
flutter checks

yohancai commented:
> flutter checks

its fixed!

JotamSleonardo commented on lib/features/favorite_list/favorite_connector.dart:
```diff
@@ -0,0 +1,20 @@
+  const FavoriteConnector({Key? key}) : super(key: key);
+
+  static const String route = '/favorites';
+  static const String name = 'favorites';
```
Are you using this for routing? if yes can you be descriptive with the naming and also I think you can reuse the existing favorites from the constants 🤔

yohancai replied:
Yes it is used for routing but I think the name is descriptive as it is once you used it (ex. PokemonListConnector.route) I removed the constants and instead made a static const of them to limit their access if the parent class is used.

JotamSleonardo replied:
@yohancai route is fine what i mean here is the name.

yohancai replied:
should i change name to pokemonName and favoriteName?

JotamSleonardo commented on lib/features/favorite_list/favorite_vm.dart:
```diff
@@ -0,0 +1,20 @@
+class FavoriteVmFactory extends VmFactory<AppState, FavoriteConnector> {
+  @override
+  Vm fromStore() {
+    return FavoriteVm(pokemons: state.pokemons);
```
you can use lambda

JotamSleonardo commented on lib/features/pokemon_list/pokemon_list_vm.dart:
```diff
@@ -5,7 +5,7 @@ import 'package:pokedex/state/app_state.dart';
   @override
   Vm fromStore() {
-    return PokemonListVm(name: state.name);
+    return PokemonListVm(pokemons: state.pokemons);
```
lambda

JotamSleonardo commented on lib/utils/generate_pokemon_image.dart:
```diff
@@ -0,0 +1,6 @@
+import 'package:pokedex/utils/constant.dart' as k;
+
+String generateImage(int index) {
```
You are not generating image here you are just getting the image url

Commit:
[PKM-30] feat: Go router setup

- changed lambda
- removed unused constants
- renamed get image url

Commit:
[PKM-30] feat: Go router setup

- change name to routeName

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,79 @@
+import 'package:pokedex/utils/strings.dart' as str;
+import 'package:pokedex/utils/constant.dart' as k;
+
+class HomePage extends StatefulWidget {
```
Do you really still need to make this stateful? 🙂

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,79 @@
+    final currentNavIndex = _selectNavIndex(router.location);
+    final navBarItems = <BottomNavigationBarItem>[
+      const BottomNavigationBarItem(
+        label: str.pokedexTitle,
```
how about if you use `'Home'` string here, coz having `'Pokedex'` is a bit misleading based on the used icon? 🙂

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,79 @@
+    final navBarItems = <BottomNavigationBarItem>[
+      const BottomNavigationBarItem(
+        label: str.pokedexTitle,
+        icon: ImageIcon(AssetImage(k.homeIcon)),
```
instead of importing new common assets for icon, would it be better if you just utilize `Icon` widget then use its varieties of predefined icons via `Icons.[anything you woud want]`? 🙂

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,79 @@
+      ),
+      const BottomNavigationBarItem(
+        label: str.favoriteTitle,
+        icon: ImageIcon(AssetImage(k.favoriteIcon)),
```
same here for the icon, see above feedback 🙂

shawn-ffufm commented on lib/state/app_state.dart:
```diff
@@ -7,8 +7,7 @@ part 'app_state.g.dart';
   factory AppState({
-    /// TODO: this is for testing purposes and will be changed once the state is finalized
-    @Default('') @JsonKey(ignore: true) String name,
+    @JsonKey(ignore: false) String? pokemons,
```
any reason why the todo comment is removed now? 🤔

shawn-ffufm commented on lib/utils/constant.dart:
```diff
@@ -1 +1,6 @@
 const double hundred = 100;
+const String favorites = 'Favorites';
+const String homeIcon = 'assets/home.png';
```
this can be removed now once you prefer to utilize the existing/built-in icons based on my feedback above 🙂

shawn-ffufm commented on lib/utils/get_image_url.dart:
```diff
@@ -0,0 +1,6 @@
+import 'package:pokedex/utils/constant.dart' as k;
+
+String getImageUrl(int index) {
```
this seems to be unused, do you still need this? 🤔

shawn-ffufm commented on lib/utils/routers.dart:
```diff
@@ -0,0 +1,38 @@
+        child: const HomePage(child: PokemonListPage()),
+        transitionsBuilder: (context, anim, _, child) => FadeTransition(opacity: anim, child: child),
+      ),
+      routes: [],
```
kindly add a todo comment that later on you will be adding sub_routes here when necessary 🙂

shawn-ffufm commented on lib/utils/routers.dart:
```diff
@@ -0,0 +1,38 @@
+        child: const HomePage(child: FavoritePage()),
+        transitionsBuilder: (context, anim, _, child) => FadeTransition(opacity: anim, child: child),
+      ),
+      routes: [],
```
same goes here 🙂

shawn-ffufm commented on pubspec.yaml:
```diff
@@ -43,3 +43,7 @@ flutter:
 
   uses-material-design: true
 
+  assets:
```
this can also be removed now once you prefer to utilize the existing/built-in icons based on my feedback above 🙂

shawn-ffufm commented on lib/utils/constant.dart:
```diff
@@ -1 +1,6 @@
 const double hundred = 100;
+const String favorites = 'Favorites';
+const String homeIcon = 'assets/home.png';
+const String favoriteIcon = 'assets/favorite.png';
```
same here 🙂

shawn-ffufm reviewed (changes requested):
@yohancai see pr feedback 🙂

shawn-ffufm reviewed (changes requested):
@yohancai also I think it would be better if you rename your child feature directories 🙈:

- Instead of `favorite_list` rename it to `pokemon_favorites` => `lib/features/pokemon_favorites`
- Instead of `pokemon_list` rename it to `pokemon_overview` => `lib/features/pokemon_overview`
- and so on later... 🙂

Commit:
[PKM-30] feat: Go router setup

- used built-in icons
- renamed directory
- made home page stateless

shawn-ffufm reviewed (changes requested):
@yohancai see pr feedback

also I have additional two points:

- kindly remove the assets you have added in your` assets directory` since those are not needed anymore 🙂 
- now that you have renamed your feature directories, kindly rename as well its components and have it correlate with the directory name 🙂

shawn-ffufm commented on lib/utils/strings.dart:
```diff
@@ -1,4 +1,3 @@
-/// TODO: this string pokemon image is temporary and dynamic data will be used
-const pokeImage = "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png";
-const String pokedexTitle = 'Pokedex';
+const String homeTitle = 'Pokedex';
```
why is that the variable name is `homeTitle` and yet its value is `Pokedex`? 😅

shawn-ffufm commented on lib/utils/constant.dart:
```diff
@@ -1 +1,4 @@
 const double hundred = 100;
+const String favorites = 'Favorites';
+const String imageURL = 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/index.png';
+const String index = 'index';
```
what's this for? 🤔

yohancai replied:
which one?

shawn-ffufm replied:
the `index` constant variable 🙂

yohancai replied:
oh its unused I have deleted it

shawn-ffufm commented on lib/state/app_state.dart:
```diff
@@ -8,7 +8,7 @@ part 'app_state.g.dart';
   factory AppState({
     /// TODO: this is for testing purposes and will be changed once the state is finalized
-    @Default('') @JsonKey(ignore: true) String name,
+    @JsonKey(ignore: true) String? pokemons,
```
Now my next question would be, is there any reason why the default annotation is removed and made this state property nullable? 🤔

yohancai replied:
there is no particular reason I just thought that it may result to null but I think I should just revert it

shawn-ffufm replied:
have it reverted then, that's the use of the `@Default` annotation to supply with a default value, meaning in this case it won't get null unless you provided a null default value 🙂

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+  String _onNavigateToTab(int selectedTabIndex, int currentTabIndex) {
```
Prefix "on" is most often used to indicate that method is intended to be used as a callback, i.e. not called directly, but set as a handler for some event.

For example, when you write method named onClick, you probably won't call it directly, but rather expect that GUI toolkit will call it once user clicks a button.

Now the thing here is that you have used the prefix `"on"` and yet you're enforcing this function to supply a return object of String? 🙈 

instead of misusing the naming you gave for this function, how about if you go for this solution and have this passed immediately on BottomNavigationBar's `onTap` parameter 🙂 :

```
void _onNavigateToTab(int selectedTabIndex, int currentTabIndex) {
    if (currentTabIndex == selectedTabIndex) return;

    final navPath = _selectNavPath(selectedTabIndex);
    context.go(navPath);
  }
```

yohancai replied:
the problem with that function is that there is no context present outside the widget build function

shawn-ffufm replied:
that's taken from my previous example when this widget was still stateful, for this instance can't you add a context param on that function if that's the case? 🙂 
also don't forget to apply trailing commas 😜

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+
+  String _onNavigateToTab(int selectedTabIndex, int currentTabIndex) {
+    if (currentTabIndex == selectedTabIndex) {
+      return _selectNavPath(currentTabIndex);
```
based on your implementation, are you sure you're always going to provide something that will trigger go_router even when you are in the same page? 🤔 

you need to understand that this may be working but the question is, is this performant? 🙂

shawn-ffufm commented on lib/features/pokemon_favorites/favorite_pokemon.dart:
```diff
@@ -0,0 +1,13 @@
+import 'package:flutter/material.dart';
+import 'package:pokedex/utils/constant.dart' as k;
+
+class FavoritePage extends StatelessWidget {
```
kindly align your dart file name with the class you have assigned here? 🙂 

`pokemon_favorites_page.dart` & `PokemonFavoritesPage` and same goes to other affected feature components (vm, connector and page) 🙂

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+
+  @override
+  Widget build(BuildContext context) {
+    final router = GoRouter.of(context);
```
kindly have this properties private 🙂

-  router
- currentNavIndex
-  navBarItems

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+  int _selectNavIndex(String routerLocation) {
+    switch (routerLocation) {
+      case PokemonListConnector.route:
+        return 0;
```
this is a magic integer, value can you store in your const.dart and provide with a better naming 🙂

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+      case PokemonListConnector.route:
+        return 0;
+      case FavoriteConnector.route:
+        return 1;
```
same here

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+      case FavoriteConnector.route:
+        return 1;
+      default:
+        return 0;
```
also here

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+
+  String _selectNavPath(int selectedTabIndex) {
+    switch (selectedTabIndex) {
+      case 0:
```
access the constants you have created based on my feedback above 🙂

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+    switch (selectedTabIndex) {
+      case 0:
+        return PokemonListConnector.route;
+      case 1:
```
same here

Commit:
[PKM-30] feat: Go router setup

- boss shawn pr feedback fix

Commit:
[PKM-30] feat: Go router setup

- files renamed

shawn-ffufm commented on lib/utils/strings.dart:
```diff
@@ -1,4 +1,2 @@
-/// TODO: this string pokemon image is temporary and dynamic data will be used
-const pokeImage = "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png";
-const String pokedexTitle = 'Pokedex';
```
I would say just keep this and add the new one for homeTitle 😅

shawn-ffufm commented on lib/utils/constant.dart:
```diff
@@ -1 +1,4 @@
 const double hundred = 100;
+const String imageURL = 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/index.png';
+const int counterZero = 0;
```
you just placed it in variable but still the same readability weight 😅 

how about 🙂 :
```
- const homeTabIndex = 0;
- const favoriteTabIndex = 0;
```

shawn-ffufm commented on lib/state/app_state.dart:
```diff
@@ -8,7 +8,7 @@ part 'app_state.g.dart';
   factory AppState({
     /// TODO: this is for testing purposes and will be changed once the state is finalized
-    @Default('') @JsonKey(ignore: true) String name,
+    @Default('') @JsonKey(ignore: true) String pokemons,
```
just revert everything to its previous state => `@Default('') @JsonKey(ignore: true) String name,` you don't need this for the simulation you made for go_router setup right? 🙂

shawn-ffufm commented on lib/pokedex_app.dart:
```diff
@@ -1,16 +1,18 @@
-      title: str.pokedexTitle,
-      home: PokemonListConnector(),
+    return MaterialApp.router(
+      title: str.homeTitle,
```
why is this `homeTitle` now? 🤔

shawn-ffufm commented on lib/home_page.dart:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+  void _onNavigateToTab(BuildContext context, int selectedTabIndex, int currentTabIndex) {
```
as I said earlier, trailing comma 😅

shawn-ffufm reviewed (changes requested):
@yohancai see last minor pr feedback 🙂

Commit:
[PKM-30] feat: Go router setup

- boss shawn feedback fix

shawn-ffufm reviewed (approved):
![image](https://user-images.githubusercontent.com/83454273/182136046-7a80772b-bdca-4691-97a8-97cb3f9fc449.png)

raelsj commented on lib/features/pokemon_favorites/pokemon_favorites_page.dart:
```diff
@@ -0,0 +1,13 @@
+  @override
+  Widget build(BuildContext context) {
+    return const Center(
+      child: Text(str.favoriteTitle),
```
1 parameter == no trailing comma needed

raelsj replied:
plus you can go ahead and use lambda

raelsj commented on lib/features/pokemon_overview/pokemon_overview_page.dart:
```diff
@@ -0,0 +1,14 @@
+  @override
+  Widget build(BuildContext context) {
+    return const Center(
+      child: Text('Pokemon'),
```
see comment above

raelsj commented on lib/home_page.dart:
```diff
@@ -0,0 +1,78 @@
+  void _onNavigateToTab(
+    BuildContext context,
+    int selectedTabIndex,
+    int currentTabIndex,
```
remove trailing comma. does it gives you lint?

raelsj commented on lib/utils/routers.dart:
```diff
@@ -0,0 +1,42 @@
+final router = GoRouter(
+  observers: [routeObservers],
+  routes: <GoRoute>[
+    GoRoute(path: '/', redirect: (_) => PokemonOverviewConnector.route, routes: []),
```
trailing comma

raelsj commented on lib/utils/routers.dart:
```diff
@@ -0,0 +1,42 @@
+      pageBuilder: (context, state) => CustomTransitionPage<void>(
+        key: state.pageKey,
+        child: const HomePage(child: PokemonOverviewPage()),
+        transitionsBuilder: (context, anim, _, child) => FadeTransition(opacity: anim, child: child),
```
trailing comma

raelsj commented on lib/utils/routers.dart:
```diff
@@ -0,0 +1,42 @@
+      pageBuilder: (context, state) => CustomTransitionPage<void>(
+        key: state.pageKey,
+        child: const HomePage(child: PokemonFavoritesPage()),
+        transitionsBuilder: (context, anim, _, child) => FadeTransition(opacity: anim, child: child),
```
trailing comma

raelsj commented on lib/utils/routers.dart:
```diff
@@ -0,0 +1,42 @@
+  observers: [routeObservers],
+  routes: <GoRoute>[
+    GoRoute(path: '/', redirect: (_) => PokemonOverviewConnector.route, routes: []),
+    GoRoute(
```
this is actually DRY. is `GoRoute` a widget?

raelsj replied:
actually never mind this comment

Commit:
[PKM-30] feat: Go router setup

- trailing commas

Commit:
[PKM-30] feat: Go router setup

- fixed boss rael feedback
